solutions/silver/usaco-895.mdx

```diff
@@ -2,234 +2,272 @@
 id: usaco-895
 source: USACO Silver 2019 January
 title: Icy Perimeter
-author: Tanish Tyagi, Brad Ma, Juheon Rhee
+author: Tanish Tyagi, Brad Ma, Juheon Rhee, David Guo
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_perimeter_silver_jan19.html)
 
+## Explanation
+
+We can use BFS to explore each blob, starting from an unvisited '#' cell.
+During BFS, we can calculate the blob’s area by counting the number of '#' cells visited and determine its perimeter by carefully analyzing the surroundings of each '#' cell.
+
+The perimeter is determined by counting the number of sides of the cell that are either adjacent to a '.' cell or are on the boundary of the grid.
+For example, if a '#' cell is surrounded on all four sides by other '#' cells, it is considered to be completely inside the blob and does not contribute to the perimeter.
+However, for each side of a '#' cell that is either adjacent to a '.' cell or is on the edge of the grid (top, bottom, left, or right boundaries), that side contributes one unit to the perimeter.
+
+As we traverse the grid, we can initiate a BFS for each unvisited '#' cell to compute the area and perimeter of all blobs.
+Within the BFS, for each '#' cell, we start by assuming that all four of its sides contribute to the perimeter and then reduce this count by one for each neighboring
+'#' cell in the grid (since shared edges do not add to the perimeter).
+
 ## Implementation
+
 **Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
+<CPPSection>
 
-<JavaSection>
-
-```java
-import java.io.*;
-import java.util.*;
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
 
-public class IcyPerimeter {
-	public static void main(String[] args) throws IOException {
-		Kattio io = new Kattio("perimeter");
-		n = io.nextInt();
-
-		// we do n + 2 because there is an extra outer ring for flood fill
-		numberedMatrix = new int[n + 2][n + 2];
-
-		for (int x = 1; x <= n; x++) {  // read matrix
-			String line = io.next();
-			for (int y = 1; y <= n; y++) {
-				String character = Character.toString(line.charAt(y - 1));
-				if (character.equals("#")) {
-					numberedMatrix[x][y] = -1;  // mark # as -1
+// Movement directions (up, down, left, right)
+const vector<pair<int, int>> DIRECTIONS = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+pair<int, int> bfs(int row, int col, vector<vector<char>> &grid,
+                   vector<vector<bool>> &visited, int n) {
```

Comment:
1. can `grid` be made a const ref
2. `n` seems redundant given that you can determine it from grid size, but idk